README.md
[![Known Vulnerabilities](https://snyk.io/test/github/stellarbeat/js-stellarbeat-backend/badge.svg)](https://snyk.io/test/github/stellarbeat/js-stellarbeat-backend)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
# Stellarbeat backend 
_Warning: Backend is under active development and subject to a changing API. Use at own risk. BC breaks between versions will be documented in changelog file._

The backend consists of three major modules.
1) Network-scan: Collects, updates and exposes network data (nodes, organizatins, stats,...). 
2) Notifications: Allows users to subscribe to network events and receive (email) notifications.
3) History-scan: Scans history archives for errors. 

The core module contains app wide functionality like logging, configuration, database, etc.

## install

````
> yarn install # install dependencies
> yarn build # build the code into the lib folder
# Database migrations are run automatically when the code is first run.
````
Copy env.dist to .env and configure the environment variables.

## Usage
Every package has a README.md file with more detailed information.

### Run the api
````
yarn start-api
````

### Run the history-scan
````
yarn verify-archives
````

### Run the network-scan
````
yarn scan-network
````

### Run tests

The test folder contains both unit and integration tests. 
A test database is required for the integration tests that you can configure with the DATABASE_TEST_URL env variable.

```` 
yarn test 
yarn test:unit
yarn test:integration
````

CHANGELOG.md
## 0.3.0
BC Breaks: 
* Network environment variable now takes the full network string.
`NETWORK="Public Global Stellar Network ; September 2015"`
* To enable Deadman snitch heartbeat, set ENABLE_HEART_BEAT to true.
* To enable Sentry exception logging, set ENABLE_SENTRY to true.
* To enable AWS backup: set ENABLE_S3_BACKUP to true.
* To be able to connect to other nodes, set the following env variables and keep them up-to-date:
  * CRAWLER_NODE_LEDGER_VERSION=18,
  * CRAWLER_NODE_OVERLAY_VERSION=18,
  * CRAWLER_NODE_OVERLAY_MIN_VERSION=17.
* The script update-nodes was renamed to update-network and can be found in src/network/infrastructure/cli. Use yarn update-network to avoid being impacted by these changes.

## 0.4.0
Network update component refactored out of network component for better component cohesion

## 0.5.0
History scan module for archive scanning and verification. Improved module naming and structure.

## 0.6.0
* Replaced TRUSTED_TOP_TIER_NODES with NETWORK_QUORUM_SET in json format 
to define the nodes that are used to determine validating status for the network. 
It is now json format to allow grouping by org. For example: ["publicKeyA", "publicKeyB", ["ORG1ValidatorA", "ORG1ValidatorB"], ["ORG2ValidatorA", "ORG2ValidatorB"]];
This will map to a quorumSet with simple majority thresholds. In the example the main threshold is 3 where publicKeyA, publicKeyB, Org1 and Org2 get a vote. 
* Loop option is moved to network update cli command parameter instead of env variable. Added dry-run option to network update cli command.
* UpdateNetwork useCase is now ScanNetwork. UpdateNetwork now only updates the network information and does not perform any measurements/crawling/....
* NETWORK env variable is now split out in 
  * NETWORK_PASSPHRASE="Public Global Stellar Network ; September 2015"
  * NETWORK_ID=public (slug of the network name)
  * NETWORK_NAME="Public Stellar Network"
* CRAWLER env settings moved to network:
  * NETWORK_OVERLAY_VERSION=24
  * NETWORK_LEDGER_VERSION=20
  * NETWORK_OVERLAY_MIN_VERSION=17
  * NETWORK_STELLAR_CORE_VERSION=19.6.0 => should match latest version of https://github.com/stellar/stellar-core
## 0.7.0
* homeDomain is now a required and unique field for organization. Migration is handled automatically. 
* SeedDb is no longer necessary, you can define KNOWN_PEERS in the .env file to bootstrap the network-scan.
* BC break: NETWORK_KNOWN_PEERS is now a mandatory env variable. 
## 0.8.0
* AWS_REGION now mandatory env variable

src/Config.ts
import { config } from 'dotenv';
config();

import { isArray, isNumber, isString } from './utilities/TypeGuards';
import { err, ok, Result } from 'neverthrow';
import * as yn from 'yn';
import { Url } from './value-objects/Url';
import { CrawlerConfiguration } from '@stellarbeat/js-stellar-node-crawler';

type PublicKey = string;

export interface Config {
	topTierFallback: PublicKey[];
	loop: boolean;
	nodeEnv: string;
	enableSentry: boolean;
	sentryDSN: string | undefined;
	ipStackAccessKey: string;
	horizonUrl: Url;
	apiCacheClearUrl: Url;
	apiCacheClearToken: string;
	deadManSwitchUrl: Url | undefined;
	enableDeadManSwitch: boolean;
	enableS3Backup: boolean;
	s3AccessKeyId: string | undefined;
	s3Secret: string | undefined;
	s3BucketName: string | undefined;
	environment: string | undefined;
	apiPort: number;
	userAgent: string;
	crawlerConfig: CrawlerConfiguration;
}

export class DefaultConfig implements Config {
	loop = false;
	nodeEnv = 'development';
	enableSentry = false;
	sentryDSN: string | undefined = undefined;
	enableDeadManSwitch = false;
	deadManSwitchUrl: Url | undefined;
	s3AccessKeyId: string | undefined;
	s3Secret: string | undefined;
	s3BucketName: string | undefined;
	enableS3Backup = false;
	environment: string | undefined;
	apiPort = 3000;
	userAgent = 'https://github.com/stellarbeat/js-stellarbeat-backend';

	constructor(
		public topTierFallback: PublicKey[],
		public horizonUrl: Url,
		public ipStackAccessKey: string,
		public apiCacheClearUrl: Url,
		public apiCacheClearToken: string,
		public crawlerConfig: CrawlerConfiguration
	) {
		this.topTierFallback = topTierFallback;
		this.horizonUrl = horizonUrl;
		this.ipStackAccessKey = ipStackAccessKey;
		this.apiCacheClearToken = apiCacheClearToken;
		this.apiCacheClearUrl = apiCacheClearUrl;
		this.crawlerConfig = crawlerConfig;
	}
}

export function getConfigFromEnv(): Result<Config, Error> {
	const topTierFallbackRaw = process.env.TOP_TIER_FALLBACK;
	if (!isString(topTierFallbackRaw))
		return err(new Error('TOP_TIER_FALLBACK not a string'));

	const topTierFallbackArray = topTierFallbackRaw.split(' ');
	if (!isArray(topTierFallbackArray))
		return err(
			new Error(
				'TOP_TIER_FALLBACK wrong format: needs space separated public keys'
			)
		);
	if (topTierFallbackArray.length === 0)
		return err(
			new Error('TOP_TIER_FALLBACK must contain at least one public key')
		);

	const ipStackAccessKey = process.env.IPSTACK_ACCESS_KEY;
	if (!isString(ipStackAccessKey))
		return err(new Error('Ipstack access key not defined'));

	const horizonUrl = process.env.HORIZON_URL;
	if (!isString(horizonUrl))
		return err(new Error('HORIZON_URL is not defined'));
	const horizonUrlResult = Url.create(horizonUrl);
	if (horizonUrlResult.isErr()) return err(horizonUrlResult.error);

	const apiCacheClearToken = process.env.BACKEND_API_CACHE_TOKEN;
	if (!isString(apiCacheClearToken))
		return err(new Error('BACKEND_API_CACHE_TOKEN not defined'));

	const apiCacheClearUrl = process.env.BACKEND_API_CACHE_URL;
	if (!isString(apiCacheClearUrl))
		return err(new Error('BACKEND_API_CACHE_URL is not defined'));
	const apiCacheClearUrlResult = Url.create(
		apiCacheClearUrl + '?token=' + apiCacheClearToken
	);
	if (apiCacheClearUrlResult.isErr()) return err(apiCacheClearUrlResult.error);

	const crawlerMaxConnectionsRaw = process.env.CRAWLER_MAX_CONNECTIONS;
	let crawlerMaxConnections = 25;
	if (isNumber(Number(crawlerMaxConnectionsRaw)))
		crawlerMaxConnections = Number(crawlerMaxConnectionsRaw);

	const crawlerNodePrivateKey = process.env.CRAWLER_NODE_PRIVATE_KEY;
	const crawlerNodeLedgerVersion = Number(
		process.env.CRAWLER_NODE_LEDGER_VERSION
	);
	const crawlerNodeOverlayVersion = Number(
		process.env.CRAWLER_NODE_OVERLAY_VERSION
	);
	const crawlerNodeOverlayMinVersion = Number(
		process.env.CRAWLER_NODE_OVERLAY_MIN_VERSION
	);
	const crawlerNodeVersionString = process.env.CRAWLER_NODE_VERSION_STRING;

	let network = process.env.NETWORK;
	if (!isString(network))
		network = 'Public Global Stellar Network ; September 2015';

	const crawlerMaxCrawlTime = Number(process.env.CRAWLER_MAX_CRAWL_TIME);

	const crawlerConfig: CrawlerConfiguration = {
		maxOpenConnections: crawlerMaxConnections,
		maxCrawlTime: Number.isNaN(crawlerMaxCrawlTime)
			? 900000
			: crawlerMaxCrawlTime,
		nodeConfig: {
			network: network,
			listeningPort: 11625,
			privateKey: crawlerNodePrivateKey,
			receiveSCPMessages: true,
			receiveTransactionMessages: false,
			nodeInfo: {
				networkId: network,
				ledgerVersion: Number.isNaN(crawlerNodeLedgerVersion)
					? 18
					: crawlerNodeLedgerVersion,
				overlayMinVersion: Number.isNaN(crawlerNodeOverlayMinVersion)
					? 17
					: crawlerNodeOverlayMinVersion,
				overlayVersion: Number.isNaN(crawlerNodeOverlayVersion)
					? 18
					: crawlerNodeOverlayVersion,
				versionString: isString(crawlerNodeVersionString)
					? crawlerNodeVersionString
					: 'sb-backend-v0.3.0'
			}
		}
	};

	const config = new DefaultConfig(
		topTierFallbackArray,
		horizonUrlResult.value,
		ipStackAccessKey,
		apiCacheClearUrlResult.value,
		apiCacheClearToken,
		crawlerConfig
	);

	const loop = yn(process.env.LOOP);
	if (loop !== undefined) config.loop = loop;

	const env = process.env.NODE_ENV;
	if (isString(env)) config.nodeEnv = env;

	const enableSentry = yn(process.env.ENABLE_SENTRY);
	config.enableSentry = enableSentry === undefined ? false : enableSentry;
	config.sentryDSN = process.env.SENTRY_DSN;

	let enableDeadManSwitch = yn(process.env.ENABLE_HEART_BEAT);
	if (enableDeadManSwitch === undefined) {
		enableDeadManSwitch = false;
	}

	config.enableDeadManSwitch = enableDeadManSwitch;
	if (config.enableDeadManSwitch) {
		const deadManSwitchUrl = process.env.DEADMAN_URL;
		if (!isString(deadManSwitchUrl))
			return err(new Error('DEADMAN_URL not defined'));
		const deadManSwitchUrlResult = Url.create(deadManSwitchUrl);
		if (deadManSwitchUrlResult.isErr())
			return err(deadManSwitchUrlResult.error);
		config.deadManSwitchUrl = deadManSwitchUrlResult.value;
	}

	let enableS3Backup = yn(process.env.ENABLE_S3_BACKUP);
	if (enableS3Backup === undefined) enableS3Backup = false;
	config.enableS3Backup = enableS3Backup;

	if (config.enableS3Backup) {
		const awsAccessKeyId = process.env.AWS_ACCESS_KEY;
		if (!isString(awsAccessKeyId))
			return err(new Error('AWS_ACCESS_KEY not defined'));
		config.s3AccessKeyId = awsAccessKeyId;

		const awsSecretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
		if (!isString(awsSecretAccessKey))
			return err(new Error('AWS_SECRET_ACCESS_KEY not defined'));
		config.s3Secret = awsSecretAccessKey;

		const awsBucketName = process.env.AWS_BUCKET_NAME;
		if (!isString(awsBucketName))
			return err(new Error('AWS_BUCKET_NAME not defined'));
		config.s3BucketName = awsBucketName;
	}

	const apiPortString = process.env.PORT;
	if (isString(apiPortString)) config.apiPort = Number(apiPortString);

	const userAgent = process.env.USER_AGENT;
	if (isString(userAgent)) config.userAgent = userAgent;

	return ok(config);
}
